pkg/linter/linter.go

```diff
@@ -421,72 +435,53 @@ func (l Linter) prepareRegoArgs() []func(*rego.Rego) {
 	return regoArgs
 }
 
-func (l Linter) lintWithRegoRules(ctx context.Context, input rules.Input) (report.Report, error) {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
-	regoArgs := l.prepareRegoArgs()
-
-	linterQuery, err := rego.New(regoArgs...).PrepareForEval(ctx)
-	if err != nil {
-		return report.Report{}, fmt.Errorf("failed preparing query for linting: %w", err)
-	}
-
+func (l Linter) lintWithRegoRules(
+	ctx context.Context, input rules.Input, aggregates []report.Aggregate,
+) (report.Report, error) {
 	aggregate := report.Report{}
 
-	var wg sync.WaitGroup
+	regoArgs := l.prepareRegoArgs(query)
 
-	var mu sync.Mutex
+	var linterQuery rego.PreparedEvalQuery
 
-	errCh := make(chan error)
-	doneCh := make(chan bool)
+	var err error
 
-	for _, name := range input.FileNames {
-		wg.Add(1)
+	if linterQuery, err = rego.New(regoArgs...).PrepareForEval(ctx); err != nil {
+		return report.Report{}, fmt.Errorf("failed preparing query for linting: %w", err)
+	}
 
-		go func(name string) {
-			defer wg.Done()
+	err = l.evalAndCollect(ctx, input, linterQuery,
 
+		func(name string, content string, module *ast.Module) (map[string]any, error) {
 			enhancedAST, err := parse.EnhanceAST(name, input.FileContent[name], input.Modules[name])
 			if err != nil {
-				errCh <- fmt.Errorf("failed preparing AST: %w", err)
-
-				return
+				return nil,
+					fmt.Errorf("could not enhance AST when building input during lint with Rego rules: %w", err)
 			}
-
-			resultSet, err := linterQuery.Eval(ctx, rego.EvalInput(enhancedAST))
-			if err != nil {
-				errCh <- fmt.Errorf("error encountered in query evaluation %w", err)
-
-				return
-			}
-
-			result, err := resultSetToReport(resultSet)
-			if err != nil {
-				errCh <- fmt.Errorf("failed to convert result set to report: %w", err)
-
-				return
+			if aggregates == nil {
+				// if the value is nil, inserting it "as is" will make it null-defined, and
+				// `count(input.aggregate) == 0` and `not input.aggregate` will not evaluate.
+				// For simplicity and to avoid error-prone code, ensure it's always defined as an array,
+				// so that authors can always check `count(input.aggregate)`
+				// not worrying about nil/null/undefined nuances.
```

Comment:
The rationales of ensuring it's always defined as part of the public contract is to avoid common mistakes (as I've seen) like this one:

1. Rego author accidentally breaks the collection of some aggregate data.
2. `input.aggregage` becomes undefined.
3. Violations using `input.aggregate[_]....` will not evaluate, and the linter will pass.

By making #2 impossible, we simplify the work/validations Rego authors need to make.